engine-sdk/src/bn128.rs

```diff
@@ -0,0 +1,518 @@
+
+    // X = LE -> BE
+    output[0..FQ_LEN].reverse();
+    // Y = LE -> BE
+    output[FQ_LEN..G1_LEN].reverse();
+    Ok(output)
+}
+
+#[cfg(not(feature = "contract"))]
+pub fn alt_bn128_g1_sum(input_bytes: &[u8]) -> Result<[u8; 64], Bn254Error> {
+    let input = utils::right_pad::<ADD_INPUT_LEN>(input_bytes);
+    let p1_bytes = &input[..G1_LEN];
+    let p2_bytes = &input[G1_LEN..];
+    utils::g1_point_add(p1_bytes, p2_bytes)
+}
+
+/// Multiplies a G1 point on the bn128 curve by a scalar via NEAR host function.
+#[cfg(feature = "contract")]
+pub fn alt_bn128_g1_scalar_multiple(input_bytes: &[u8]) -> Result<[u8; 64], Bn254Error> {
+    // Buffer is: [P1(G1_LEN), Scalar(SCALAR_LEN)] -> Total 96 bytes
+    const BUFFER_LEN: usize = G1_LEN + SCALAR_LEN;
+    // Register ID to store the result
+    const REGISTER_ID: u64 = 1;
+
+    let mut bytes = [0u8; BUFFER_LEN];
+
+    // 1. Process Point G1 (First 64 bytes)
+    // X coordinate (Input 0..32)
+    write_reversed_chunk(&mut bytes[0..FQ_LEN], input_bytes, 0);
+    // Y coordinate (Input 32..64)
+    write_reversed_chunk(&mut bytes[FQ_LEN..G1_LEN], input_bytes, FQ_LEN);
+
+    // 2. Process Scalar (Next 32 bytes)
+    // Scalar (Input 64..96)
+    write_reversed_chunk(&mut bytes[G1_LEN..], input_bytes, G1_LEN);
+
+    let value_ptr = bytes.as_ptr() as u64;
+    let value_len = bytes.len() as u64;
+    // Prepare output buffer
+    let mut output = [0u8; G1_LEN];
+    // Call the NEAR host function
+    unsafe {
+        exports::alt_bn128_g1_multiexp(value_len, value_ptr, REGISTER_ID);
+        exports::read_register(REGISTER_ID, output.as_ptr() as u64);
+    }
+
+    // 3. Output Point G1 (Little -> Big)
+    // X coordinate
+    output[0..FQ_LEN].reverse();
+    // Y coordinate
+    output[FQ_LEN..G1_LEN].reverse();
+
+    Ok(output)
+}
+
+#[cfg(not(feature = "contract"))]
+pub fn alt_bn128_g1_scalar_multiple(input_bytes: &[u8]) -> Result<[u8; 64], Bn254Error> {
+    let input = utils::right_pad::<MUL_INPUT_LEN>(input_bytes);
+
+    let point_bytes = &input[..G1_LEN];
+    let scalar_bytes = &input[G1_LEN..G1_LEN + SCALAR_LEN];
+    utils::g1_point_mul(point_bytes, scalar_bytes)
+}
+
+/// Performs a pairing check on a list of G1 and G2 point pairs via NEAR host function.
+/// Accepts a byte slice containing a sequence of pairs (G1, G2).
+#[cfg(feature = "contract")]
+pub fn alt_bn128_pairing(input_bytes: &[u8]) -> Result<bool, Bn254Error> {
+    use aurora_engine_types::Vec;
+
+    // Empty input implies the product of an empty set, which is the multiplicative identity (1).
+    // Therefore, the check passes.
+    if input_bytes.is_empty() {
+        return Ok(true);
+    }
+
+    // Validate input length
+    if input_bytes.len() % PAIR_ELEMENT_LEN != 0 {
+        return Err(Bn254Error::InvalidPairLength);
+    }
+
+    let len = input_bytes.len();
+    let mut bytes = Vec::with_capacity(len);
+
+    // Unsafe Direct Write - bypass Vectors safety checks (extend_from_slice) to save gas.
```

Comment:
I'm curious how much gas this saves. Working with raw pointers is dangerous, but could be justifiable if the gas savings are high enough.

Reply:
Could you please add one large input test to check the gas savings? No need to do the whole test suite parsing as part of this PR; I understand that will be a lot of valuable work for the future, but I would like to see a measurement of the gas savings here as a justification for using this optimization. I think we should have a relatively high bar for introducing unsafe code into our code base since we are dealing with security-critical software.